cluster_management/src/main/java/com/pinterest/rocksplicator/eventstore/LeaderEventHistoryStore.java

```diff
@@ -0,0 +1,499 @@
+/// You may obtain a copy of the License at
+///
+/// http://www.apache.org/licenses/LICENSE-2.0
+
+/// Unless required by applicable law or agreed to in writing, software
+/// distributed under the License is distributed on an "AS IS" BASIS,
+/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+/// See the License for the specific language governing permissions and
+/// limitations under the License.
+
+//
+// @author Gopal Rajpurohit (grajpurohit@pinterest.com)
+//
+
+package com.pinterest.rocksplicator.eventstore;
+
+import static com.pinterest.rocksplicator.thrift.eventhistory.LeaderEventType.CLIENT_OBSERVED_SHARDMAP_LEADER_DOWN;
+import static com.pinterest.rocksplicator.thrift.eventhistory.LeaderEventType.CLIENT_OBSERVED_SHARDMAP_LEADER_UP;
+import static com.pinterest.rocksplicator.thrift.eventhistory.LeaderEventType.SPECTATOR_OBSERVED_LEADER_DOWN;
+import static com.pinterest.rocksplicator.thrift.eventhistory.LeaderEventType.SPECTATOR_OBSERVED_LEADER_UP;
+import static com.pinterest.rocksplicator.thrift.eventhistory.LeaderEventType.SPECTATOR_POSTED_SHARDMAP_LEADER_DOWN;
+import static com.pinterest.rocksplicator.thrift.eventhistory.LeaderEventType.SPECTATOR_POSTED_SHARDMAP_LEADER_UP;
+
+import com.pinterest.rocksplicator.codecs.Codec;
+import com.pinterest.rocksplicator.codecs.WrappedDataThriftCodec;
+import com.pinterest.rocksplicator.thrift.commons.io.CompressionAlgorithm;
+import com.pinterest.rocksplicator.thrift.commons.io.SerializationProtocol;
+import com.pinterest.rocksplicator.thrift.eventhistory.LeaderEvent;
+import com.pinterest.rocksplicator.thrift.eventhistory.LeaderEventType;
+import com.pinterest.rocksplicator.thrift.eventhistory.LeaderEventsHistory;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheLoader;
+import com.google.common.cache.LoadingCache;
+import com.google.common.cache.RemovalListener;
+import com.google.common.cache.RemovalNotification;
+import com.google.common.math.IntMath;
+import org.apache.curator.framework.CuratorFramework;
+import org.apache.curator.framework.CuratorFrameworkFactory;
+import org.apache.curator.retry.ExponentialBackoffRetry;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.function.BiConsumer;
+
+public class LeaderEventHistoryStore implements Closeable {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(LeaderEventHistoryStore.class);
+  private static final int NUM_PARALLEL_THREADS = 64;
```

Comment:
I will run in both Participant as well as Spectator... Participant doesn't have default 40 threads. That is helix's default limit on number of threads used for callback to statemodels for state transitions. Here the LeaderEventHistoryStore maintains it's own executor service and provides async way to merge leader events to Participant (state transitions). So it will create it's own threads. Moreover we don't want to block participant state transitions while we merge data to zk for leader events history.